File: 구/jamiehun/2992.py
# 오답 => 못풀었음

x = input()

number = ""
minNum = "999999"
digit = len(x)
used = [False] * digit

# 최대자릿수는 6자리이므로 최대 재귀 깊이도 6이 됨
def backtrack(depth):
    global number
    global minNum
    
    if depth == digit:
        # 주어진 숫자 < number < minNum가 답임
        if x < number < minNum:
            minNum = number
        return
    
    for i in range(digit):
        if used[i] == True: # if i in used: (BAD)
            continue
        used[i] = True 
        number += x[i] 
        backtrack(depth + 1)
        used[i] = False 
        number = number[:-1]
        
backtrack(0)

if minNum == '999999':
    minNum = 0
print(minNum)

File: 구/YoungwooKim/2992.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim();

const solution = () => {
    let number = '';
    let minNumber = '999999';
    let digit = input.length;
    let visited = Array.from({ length: digit }, () => false);

    const backtracking = (depth) => {
        if (depth == digit) {
            if (input < number && number < minNumber) minNumber = number;
            return;
        }
        for (let i = 0; i < digit; i++) {
            if (visited[i]) continue;

            visited[i] = true;
            number += input[i];
            backtracking(depth + 1);
            visited[i] = false;
            number = number.slice(0, -1);
        }
    };
    backtracking(0);
    if (minNumber === '999999') minNumber = '0';
    console.log(minNumber);
};

solution();